secure-api-gateway-ig-extensions/src/main/java/com/forgerock/sapi/gateway/jwks/sign/DefaultSapiJwsSigner.java

```diff
@@ -0,0 +1,140 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.forgerock.sapi.gateway.jwks.sign;
+
+import static org.forgerock.openig.secrets.SecretsProviderHeaplet.secretsProvider;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import org.forgerock.json.jose.builders.JwtBuilderFactory;
+import org.forgerock.json.jose.builders.SignedJwtBuilderImpl;
+import org.forgerock.json.jose.jws.JwsAlgorithm;
+import org.forgerock.json.jose.jws.SignedJwt;
+import org.forgerock.json.jose.jws.SigningManager;
+import org.forgerock.json.jose.jws.handlers.SigningHandler;
+import org.forgerock.json.jose.jwt.JwtClaimsSet;
+import org.forgerock.openig.heap.GenericHeaplet;
+import org.forgerock.openig.heap.HeapException;
+import org.forgerock.secrets.Purpose;
+import org.forgerock.secrets.SecretsProvider;
+import org.forgerock.secrets.keys.SigningKey;
+import org.forgerock.util.Reject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Default implementation of {@link SapiJwsSigner}
+ * <p>
+ * This default JWS signer is configured in the IG configuration Heap to be used in filters<br/>
+ * @see DefaultSapiJwsSigner.Heaplet
+ */
+public class DefaultSapiJwsSigner implements SapiJwsSigner {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultSapiJwsSigner.class);
+    private final SigningManager signingManager;
+    private final Purpose<SigningKey> signingKeyPurpose;
+    private final String algorithm;
+    private final String kid;
+    private SecretsProvider secretsProvider;
+    private SigningKey signingKey;
+
+    public DefaultSapiJwsSigner(
+            SecretsProvider secretsProvider,
+            String signingKeyId,
+            String kid,
+            String algorithm
+    ) {
+        this.secretsProvider = secretsProvider;
+        this.signingManager = new SigningManager(secretsProvider);
+        this.kid = kid;
+        this.algorithm = algorithm;
+        this.signingKeyPurpose = Purpose.purpose(signingKeyId, SigningKey.class);
+
+    }
+
+    @Override
+    public String sign(
+            final Map<String, Object> payload,
+            final Map<String, Object> criticalHeaderClaims
+    ) throws SapiJwsSignerException {
+        try {
+            final JwtClaimsSet jwtClaimsSet = new JwtClaimsSet(payload);
+            secretsProvider.getActiveSecret(signingKeyPurpose).then(signingKey -> this.signingKey = signingKey);
```

Comment:
The issue with this current line is that it will only work if the `then` callback is executed by the same thread as the thread executing the DefaultSapiJwsSigner.sign method. When you run your JUnit tests you'll see that this is the case, the main thread will execute the logic sequentially, so the signingKey field is getting set and you're able to use it in the subsequent code.

However, this is not guaranteed to be the case, the Promise.then callback may not be called until later by a different thread if the call is truely asynchronous. For instance if the SecretsProvider needs to fetch data from Google Secrets manager then this call will definitely be async as a non-block network call needs to be made. In which case your signing logic will fail with an NPE as the signingKey will not be set the first time it is called.

Therefore, whenever you start down the route of using Promises, then all dependent code will need to be chained in Promises too. The DefaultSapiJwsSigner.sign method will need to return a Promise itself that the caller will then need to handle.